### ISSUE TITLE
Focus taken away when dock was hidden( not respect original activeElement ).

### ISSUE THREAD

**Issue by t9md** [atom/atom#14338]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

When dock was hidden, it **force** focus on workspaceCenter's active pane( Ignore original focus ).

https://github.com/atom/atom/pull/14143/commits/73f04fdaed30b774d1f5134a1df05eba30202aab

So hiding dock means "focus is taken away from originally focused element".
I want hiding dock not cause focus change unless hidden dock was originally focused.

When package-author try to implement action which hide dock and want to keep focus on `select-list` or mini-editor.
It's unnecessarily difficult since hiding dock force focus to workspaceCenter's active pane.

- Actual situation

In my [project-folder](https://atom.io/packages/project-folder) package, user can add/remove multiple project with keeping focus on select-list.
The `space` key in `select-list` is keymapped as "confirm and continue" action which confirm current item on select-list and keep focus on `select-list` so that user can continuously remove/add project folder.
From atom v1.17(which introduced dock), when project paths get emptied, tree-view destroy itself and tree-view's dock is hidden.
At this point, **focus is taken away from select-list to workspaceCenter's active pane** and `blur` event is fired on `select-list`'s element. which cause `cancel()` of select-list.
I don't want `cancel`, or `blur`, just want focus kept on originally focused `selec-list`'s mini-editor.

This is new issue from atom v1.17.0. Was not problem in v1.16.0.

https://github.com/t9md/atom-project-folder/issues/9#issuecomment-298679687

### Steps to Reproduce

1. Set following commands and keymap.

- `init.coffee`

```coffeescript
atom.commands.add 'atom-workspace',
  'user:focus-taken-away-repro': ->
    for dir in atom.project.getPaths()
      atom.project.removePath(dir)
```

- `keymap.cson`

```coffeescript
'.select-list':
  'space': 'user:focus-taken-away-repro'
```

2. Start atom-beta(v1.17) with at least one project-folder listed on tree-view.
3. `cmd-t` to start `fuzzy-finder`.
4. type `space`, which invoke `user:focus-taken-away-repro` comands.

- In atom 1.16.0: focus remain in `select-list`'s mini-editor.
- In atom 1.17.0-beta3: focus taken away from `select-list` to activePane.

**Expected behavior:** [What you expect to happen]

Want atom not take away original focus when dock was hidden.
Want keep focused on `select-list`'s mini-editor.

**Actual behavior:** [What actually happens]

Original focus was taken away from `select-list` to activePane.

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%, allways.

### Versions

```coffeescript
Atom    : 1.17.0-beta3
Electron: 1.3.14
Chrome  : 52.0.2743.82
Node    : 6.5.0
```

**Comment by maxbrunsfeld** [comment]
Thanks for investigating this so thoroughly @t9md. I totally agree with your expected behavior.

**Comment by t9md** [comment]
Oh, thanks for quick response!

**Comment by t9md** [comment]
@maxbrunsfeld 
Thanks you for quick fixing.
I checked with https://github.com/atom/atom/pull/14339/commits.

```
Atom    : 1.18.0-dev-a61accf17
Electron: 1.3.15
Chrome  : 52.0.2743.82
Node    : 6.5.0
```

Result was partially fixed.

With the reproduce procedure I explained .

Case-1: Fixed: If I focus editor first then `cmd-t space`, focus not taken away, select-list still have focus.
Case-2: Not Fixed: If I focus tree-view first then `cmd-t space`, focus taken away, select-list closed.

For case-2 to fix, I locally changed like following.

```javascript
didHideDock (dock) {
  if (dock.element === document.activeElement || dock.element.contains(document.activeElement)) {
    if (dock === this.activePaneContainer) {
      this.getCenter().activate()
    }
  }
}
```

This fix case-2. This is how I want. I want control focus by myself. Want automatic focus-change as minimum as possible.  
But I'm also not sure this is OK.  
It might introduce big behavior diff whether command was invoked from keymap or comand-palette.  
When command was invoked via command-palette focus is on mini-editor of select-list.( But at the timing of confirm it refocus to originally focused element so might not problem. )
